<a href="https://colab.research.google.com/github/fidanaydamirova/youtube_webscraping/blob/main/Youtube_comments_scraping.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ***Introduction: We are performing webscraping of youtube video comments, storing them in a dataset and performing sentiment analysis on them to find the percentage of positive and negative reviews.***

---






In [1]:
!pip install time
!pip install nltk
!pip install selenium

ERROR: Could not find a version that satisfies the requirement time (from versions: none)
ERROR: No matching distribution found for time
     |████████████████████████████████| 958 kB 5.3 MB/s 
     |████████████████████████████████| 356 kB 45.4 MB/s 
     |████████████████████████████████| 138 kB 53.0 MB/s 
     |████████████████████████████████| 55 kB 2.7 MB/s 
     |████████████████████████████████| 3.6 MB 31.4 MB/s 
     |████████████████████████████████| 54 kB 2.4 MB/s 
  Attempting uninstall: urllib3
    Found existing installation: urllib3 1.24.3
    Uninstalling urllib3-1.24.3:
      Successfully uninstalled urllib3-1.24.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
requests 2.23.0 requires urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1, but you have urllib3 1.26.7 which is incompatible.
datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 

In [2]:
!apt install chromium-chromedriver
!cp /usr/lib/chromium-browser/chromedriver /usr/bin

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  chromium-browser chromium-browser-l10n chromium-codecs-ffmpeg-extra
Suggested packages:
  webaccounts-chromium-extension unity-chromium-extension
The following NEW packages will be installed:
  chromium-browser chromium-browser-l10n chromium-chromedriver
  chromium-codecs-ffmpeg-extra
0 upgraded, 4 newly installed, 0 to remove and 37 not upgraded.
Need to get 94.0 MB of archives.
After this operation, 324 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 chromium-codecs-ffmpeg-extra amd64 95.0.4638.69-0ubuntu0.18.04.1 [1,135 kB]
Get:2 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 chromium-browser amd64 95.0.4638.69-0ubuntu0.18.04.1 [83.6 MB]
Get:3 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 chromium-browser-l10n all 95.0.4638.69-0ubuntu0.18.0

In [3]:
#import all required libraries
import nltk
from nltk.sentiment.vader import SentimentIntensityAnalyzer
import time
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.service import Service
import pandas as pd
from statistics import mean

/usr/local/lib/python3.7/dist-packages/nltk/twitter/__init__.py:20: UserWarning: The twython library has not been installed. Some functionality from the twitter package will not be available.
  warnings.warn("The twython library has not been installed. "


In [4]:
# set options to be headless, ..
options = webdriver.ChromeOptions()
options.add_argument('--headless')
options.add_argument('--no-sandbox')
options.add_argument('--disable-dev-shm-usage')


In [5]:
data = []
# open it, go to a website, and get results
driver = webdriver.Chrome('chromedriver',options=options)
driver.get("https://www.youtube.com/watch?v=CtXEcBz3Go0&t=1s")
# scroll till the end of the video so comments will start loading
driver.execute_script('window.scrollTo(1, 500);')
# now wait let load the comments
time.sleep(5)
with driver:
    wait = WebDriverWait(driver, 5)

    #we can increase the range to get more comments
    for item in range(100):
        wait.until(EC.visibility_of_element_located((By.TAG_NAME, "body"))).send_keys(Keys.END)
        time.sleep(5)

    for comment in wait.until(EC.presence_of_all_elements_located((By.CSS_SELECTOR, "#content"))):
        data.append(comment.text)

df = pd.DataFrame(data, columns=['comment'])

In [6]:
#take a look at the top of the data set
df.head(10)

comment
0  SKIP NAVIGATION\nSIGN IN\n0:01 / 1:42\n#Extrem...
1                                                   
2                                                   
3                                                   
4  Check out the official Extremely Wicked, Shock...
5  Life wasn’t so easy for Troy Bolton after grad...
6  For everyone that's upset about the romanticiz...
7  Everyone who is saying this movie romanticized...
8  Oh god when the little girl went “ted!” I was ...
9  I don't get the hate...it's seems like a good ...

In [7]:
#drop top rows that don't contain comments
df = df.drop(df.index[range(4)])
df.head()

comment
4  Check out the official Extremely Wicked, Shock...
5  Life wasn’t so easy for Troy Bolton after grad...
6  For everyone that's upset about the romanticiz...
7  Everyone who is saying this movie romanticized...
8  Oh god when the little girl went “ted!” I was ...

In [8]:
#check how many comments we have
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1401 entries, 4 to 1404
Data columns (total 1 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   comment  1401 non-null   object
dtypes: object(1)
memory usage: 21.9+ KB


In [9]:
nltk.downloader.download('vader_lexicon')
# Instantiate the sentiment intensity analyzer
vader = SentimentIntensityAnalyzer()

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...


In [10]:
#define new column "scores" that contains the info of positive, 
#negative and neutral proportion of a comment lines
df['scores'] = df['comment'].apply(lambda comment: vader.polarity_scores(comment))

#look at the top of the data
df.head()

comment                                             scores
4  Check out the official Extremely Wicked, Shock...  {'neg': 0.538, 'neu': 0.462, 'pos': 0.0, 'comp...
5  Life wasn’t so easy for Troy Bolton after grad...  {'neg': 0.0, 'neu': 0.681, 'pos': 0.319, 'comp...
6  For everyone that's upset about the romanticiz...  {'neg': 0.16, 'neu': 0.662, 'pos': 0.178, 'com...
7  Everyone who is saying this movie romanticized...  {'neg': 0.049, 'neu': 0.818, 'pos': 0.133, 'co...
8  Oh god when the little girl went “ted!” I was ...  {'neg': 0.189, 'neu': 0.541, 'pos': 0.27, 'com...

In [11]:
# add new column "compound" which is the sum of positive, negative & neutral 
# scores which is then normalized between -1(most extreme negative) 
# and +1 (most extreme positive)
df['compound']  = df['scores'].apply(lambda score_dict: score_dict['compound'])

df.head()

comment  ... compound
4  Check out the official Extremely Wicked, Shock...  ...  -0.9837
5  Life wasn’t so easy for Troy Bolton after grad...  ...   0.5777
6  For everyone that's upset about the romanticiz...  ...   0.2332
7  Everyone who is saying this movie romanticized...  ...   0.4767
8  Oh god when the little girl went “ted!” I was ...  ...   0.4406

[5 rows x 3 columns]

In [12]:
# check if the overall comment negative or positive
df['comp_score'] = df['compound'].apply(lambda c: 'pos' if c >=0 else 'neg')

df.head()

comment  ... comp_score
4  Check out the official Extremely Wicked, Shock...  ...        neg
5  Life wasn’t so easy for Troy Bolton after grad...  ...        pos
6  For everyone that's upset about the romanticiz...  ...        pos
7  Everyone who is saying this movie romanticized...  ...        pos
8  Oh god when the little girl went “ted!” I was ...  ...        pos

[5 rows x 4 columns]

In [13]:
# here we can explore some examples from our data set and compare the results 
# which scores are consireder positive and which negative
for i in range(6,20):
    print(vader.polarity_scores(df["comment"][i]))

{'neg': 0.16, 'neu': 0.662, 'pos': 0.178, 'compound': 0.2332}
{'neg': 0.049, 'neu': 0.818, 'pos': 0.133, 'compound': 0.4767}
{'neg': 0.189, 'neu': 0.541, 'pos': 0.27, 'compound': 0.4406}
{'neg': 0.0, 'neu': 0.76, 'pos': 0.24, 'compound': 0.802}
{'neg': 0.0, 'neu': 0.769, 'pos': 0.231, 'compound': 0.6901}
{'neg': 0.307, 'neu': 0.526, 'pos': 0.167, 'compound': -0.3818}
{'neg': 0.05, 'neu': 0.782, 'pos': 0.168, 'compound': 0.7394}
{'neg': 0.078, 'neu': 0.819, 'pos': 0.102, 'compound': 0.579}
{'neg': 0.259, 'neu': 0.569, 'pos': 0.173, 'compound': -0.6597}
{'neg': 0.085, 'neu': 0.712, 'pos': 0.203, 'compound': 0.9062}
{'neg': 0.155, 'neu': 0.726, 'pos': 0.119, 'compound': -0.754}
{'neg': 0.0, 'neu': 0.645, 'pos': 0.355, 'compound': 0.8874}
{'neg': 0.081, 'neu': 0.699, 'pos': 0.22, 'compound': 0.8212}
{'neg': 0.097, 'neu': 0.702, 'pos': 0.201, 'compound': 0.9046}


In [14]:
# get the mean of positive comments
mean(df["comp_score"] == "pos")

0.7366167023554604

In [15]:
# get the mean of negative comments
mean(df["comp_score"] == "neg")

0.2633832976445396

In [16]:
import matplotlib.pyplot as plt

# Conclusion: Overall we can conclude that this particular video has 73.7% positive and 26.3% negative comments.